File: packages/sdk-ts/src/anomaly/detector.ts
/**
 * AgentOps SDK - Anomaly Detector
 *
 * ML-powered detection of unusual patterns in AI metrics.
 */

import type {
  Anomaly,
  AnomalyType,
  AnomalySeverity,
  AnomalyDetectionConfig,
  AnomalyMetricConfig,
  MetricSnapshot,
  MetricTimeSeries,
  CustomDetectionRule,
} from "./types.js";
import { generateEventId, now } from "../utils.js";

const DEFAULT_CONFIG: Required<
  Omit<AnomalyDetectionConfig, "onAnomaly" | "customRules">
> & {
  customRules: CustomDetectionRule[];
} = {
  enabled: false,
  sensitivity: 0.5,
  minDataPoints: 30,
  baselineWindowMinutes: 60,
  metrics: [
    { name: "costPerHour", method: "zscore", threshold: 3, direction: "above" },
    {
      name: "latencyP95",
      method: "zscore",
      threshold: 2.5,
      direction: "above",
    },
    {
      name: "errorRate",
      method: "threshold",
      fixedThreshold: 0.05,
      direction: "above",
    },
    {
      name: "qualityScore",
      method: "zscore",
      threshold: 2,
      direction: "below",
    },
  ],
  customRules: [],
};

interface MetricBuffer {
  values: number[];
  timestamps: number[];
  maxSize: number;
}

export class AnomalyDetector {
  private readonly config: typeof DEFAULT_CONFIG & {
    onAnomaly?: (anomaly: Anomaly) => void;
  };
  private metricBuffers: Map<string, MetricBuffer> = new Map();
  private activeAnomalies: Map<string, Anomaly> = new Map();
  private anomalyHistory: Anomaly[] = [];

  constructor(config?: AnomalyDetectionConfig) {
    this.config = {
      ...DEFAULT_CONFIG,
      ...config,
      metrics: config?.metrics ?? DEFAULT_CONFIG.metrics,
      customRules: config?.customRules ?? [],
    };
  }

  get isEnabled(): boolean {
    return this.config.enabled;
  }

  /**
   * Record a metric snapshot for analysis
   */
  recordMetrics(snapshot: MetricSnapshot): Anomaly[] {
    if (!this.config.enabled) return [];

    const detectedAnomalies: Anomaly[] = [];

    // Update buffers with new metrics
    for (const [key, value] of Object.entries(snapshot)) {
      if (typeof value === "number" && key !== "timestamp") {
        this.addToBuffer(key, value, snapshot.timestamp);
      }
    }

    // Handle custom metrics
    if (snapshot.custom) {
      for (const [key, value] of Object.entries(snapshot.custom)) {
        this.addToBuffer(`custom.${key}`, value, snapshot.timestamp);
      }
    }

    // Check configured metrics
    for (const metricConfig of this.config.metrics) {
      const buffer = this.metricBuffers.get(metricConfig.name);
      if (!buffer || buffer.values.length < this.config.minDataPoints) {
        continue;
      }

      const value = snapshot[metricConfig.name as keyof MetricSnapshot];
      if (typeof value !== "number") continue;

      const anomaly = this.detectAnomaly(
        metricConfig,
        value,
        buffer,
        snapshot.timestamp,
      );
      if (anomaly) {
        detectedAnomalies.push(anomaly);
      }
    }

    // Check custom rules
    for (const rule of this.config.customRules) {
      if (rule.condition(snapshot)) {
        const anomaly = this.createCustomAnomaly(rule, snapshot);
        detectedAnomalies.push(anomaly);
      }
    }

    // Process detected anomalies
    for (const anomaly of detectedAnomalies) {
      this.processAnomaly(anomaly);
    }

    return detectedAnomalies;
  }

  /**
   * Get currently active anomalies
   */
  getActiveAnomalies(): Anomaly[] {
    return Array.from(this.activeAnomalies.values());
  }

  /**
   * Get anomaly history
   */
  getHistory(limit?: number): Anomaly[] {
    const history = [...this.anomalyHistory].reverse();
    return limit ? history.slice(0, limit) : history;
  }

  /**
   * Resolve an anomaly
   */
  resolveAnomaly(anomalyId: string): boolean {
    const anomaly = this.activeAnomalies.get(anomalyId);
    if (!anomaly) return false;

    anomaly.isActive = false;
    anomaly.endTime = now();
    this.activeAnomalies.delete(anomalyId);

    return true;
  }

  /**
   * Get time series data for a metric
   */
  getMetricTimeSeries(metric: string): MetricTimeSeries | null {
    const buffer = this.metricBuffers.get(metric);
    if (!buffer || buffer.values.length === 0) return null;

    const dataPoints = buffer.values.map((value, i) => ({
      timestamp: buffer.timestamps[i],
      value,
    }));

    const stats = this.calculateStats(buffer.values);

    return {
      metric,
      dataPoints,
      stats,
    };
  }

  /**
   * Add a custom detection rule
   */
  addCustomRule(rule: CustomDetectionRule): void {
    this.config.customRules.push(rule);
  }

  /**
   * Remove a custom detection rule
   */
  removeCustomRule(ruleId: string): boolean {
    const index = this.config.customRules.findIndex((r) => r.id === ruleId);
    if (index === -1) return false;
    this.config.customRules.splice(index, 1);
    return true;
  }

  private addToBuffer(metric: string, value: number, timestamp: number): void {
    let buffer = this.metricBuffers.get(metric);

    if (!buffer) {
      buffer = {
        values: [],
        timestamps: [],
        maxSize: this.config.baselineWindowMinutes * 60, // Assume 1 sample/second
      };
      this.metricBuffers.set(metric, buffer);
    }

    buffer.values.push(value);
    buffer.timestamps.push(timestamp);

    // Trim old data
    const cutoff = timestamp - this.config.baselineWindowMinutes * 60 * 1000;
    while (buffer.timestamps.length > 0 && buffer.timestamps[0] < cutoff) {
      buffer.values.shift();
      buffer.timestamps.shift();
    }
  }

  private detectAnomaly(
    config: AnomalyMetricConfig,
    value: number,
    buffer: MetricBuffer,
    timestamp: number,
  ): Anomaly | null {
    const stats = this.calculateStats(buffer.values);
    let isAnomaly = false;
    let deviation = 0;
    let expectedValue = stats.mean;

    switch (config.method) {
      case "zscore": {
        const zscore =
          stats.stdDev > 0 ? (value - stats.mean) / stats.stdDev : 0;
        deviation = Math.abs(zscore);
        const threshold =
          (config.threshold ?? 3) * (1 - this.config.sensitivity * 0.5);

        if (config.direction === "above" && zscore > threshold) {
          isAnomaly = true;
        } else if (config.direction === "below" && zscore < -threshold) {
          isAnomaly = true;
        } else if (
          config.direction === "both" &&
          Math.abs(zscore) > threshold
        ) {
          isAnomaly = true;
        }
        break;
      }

      case "mad": {
        const mad = this.calculateMAD(buffer.values, stats.median);
        const modifiedZscore =
          mad > 0 ? (0.6745 * (value - stats.median)) / mad : 0;
        deviation = Math.abs(modifiedZscore);
        expectedValue = stats.median;
        const threshold =
          (config.threshold ?? 3.5) * (1 - this.config.sensitivity * 0.5);

        if (config.direction === "above" && modifiedZscore > threshold) {
          isAnomaly = true;
        } else if (
          config.direction === "below" &&
          modifiedZscore < -threshold
        ) {
          isAnomaly = true;
        } else if (
          config.direction === "both" &&
          Math.abs(modifiedZscore) > threshold
        ) {
          isAnomaly = true;
        }
        break;
      }

      case "percentile": {
        const percentile = config.percentile ?? 99;
        const threshold = this.calculatePercentile(buffer.values, percentile);
        expectedValue = threshold;
        deviation =
          value > threshold ? ((value - threshold) / threshold) * 100 : 0;

        if (config.direction === "above" && value > threshold) {
          isAnomaly = true;
        } else if (
          config.direction === "below" &&
          value < this.calculatePercentile(buffer.values, 100 - percentile)
        ) {
          isAnomaly = true;
        }
        break;
      }

      case "threshold": {
        const fixedThreshold = config.fixedThreshold ?? 0;
        expectedValue = fixedThreshold;
        deviation =
          fixedThreshold > 0
            ? (Math.abs(value - fixedThreshold) / fixedThreshold) * 100
            : 0;

        if (config.direction === "above" && value > fixedThreshold) {
          isAnomaly = true;
        } else if (config.direction === "below" && value < fixedThreshold) {
          isAnomaly = true;
        }
        break;
      }
    }

    if (!isAnomaly) return null;

    const type = this.metricToAnomalyType(config.name);
    const severity = this.calculateSeverity(deviation, config);
    const confidence = this.calculateConfidence(
      buffer.values.length,
      deviation,
    );

    return {
      id: generateEventId(),
      type,
      severity,
      title: `${type.replace(/_/g, " ").replace(/\b\w/g, (c) => c.toUpperCase())} Detected`,
      description: `${config.name} is ${value.toFixed(2)} (expected: ${expectedValue.toFixed(2)}, deviation: ${deviation.toFixed(1)}σ)`,
      metric: config.name,
      expectedValue,
      actualValue: value,
      deviation,
      confidence,
      detectedAt: timestamp,
      startTime: timestamp,
      isActive: true,
      suggestions: this.generateSuggestions(type, config.name),
    };
  }

  private createCustomAnomaly(
    rule: CustomDetectionRule,
    snapshot: MetricSnapshot,
  ): Anomaly {
    return {
      id: generateEventId(),
      type: "custom",
      severity: rule.severity,
      title: rule.name,
      description: rule.describe(snapshot),
      metric: "custom",
      expectedValue: 0,
      actualValue: 0,
      deviation: 0,
      confidence: 1,
      detectedAt: snapshot.timestamp,
      startTime: snapshot.timestamp,
      isActive: true,
      metadata: { ruleId: rule.id },
    };
  }

  private processAnomaly(anomaly: Anomaly): void {
    // Check for existing similar anomaly
    const existingKey = `${anomaly.type}:${anomaly.metric}`;
    const existing = this.activeAnomalies.get(existingKey);

    if (existing) {
      // Update existing anomaly
      existing.actualValue = anomaly.actualValue;
      existing.deviation = Math.max(existing.deviation, anomaly.deviation);
      if (
        anomaly.severity === "critical" ||
        (anomaly.severity === "warning" && existing.severity === "info")
      ) {
        existing.severity = anomaly.severity;
      }
    } else {
      // New anomaly
      this.activeAnomalies.set(existingKey, anomaly);
      this.anomalyHistory.push(anomaly);

      if (this.config.onAnomaly) {
        this.config.onAnomaly(anomaly);
      }
    }
  }

  private calculateStats(values: number[]): {
    mean: number;
    stdDev: number;
    min: number;
    max: number;
    median: number;
    p95: number;
    p99: number;
  } {
    if (values.length === 0) {
      return { mean: 0, stdDev: 0, min: 0, max: 0, median: 0, p95: 0, p99: 0 };
    }

    const sorted = [...values].sort((a, b) => a - b);
    const mean = values.reduce((a, b) => a + b, 0) / values.length;
    const variance =
      values.reduce((sum, v) => sum + Math.pow(v - mean, 2), 0) / values.length;

    return {
      mean,
      stdDev: Math.sqrt(variance),
      min: sorted[0],
      max: sorted[sorted.length - 1],
      median: this.calculatePercentile(sorted, 50),
      p95: this.calculatePercentile(sorted, 95),
      p99: this.calculatePercentile(sorted, 99),
    };
  }

  private calculatePercentile(
    sortedValues: number[],
    percentile: number,
  ): number {
    if (sortedValues.length === 0) return 0;
    const index = (percentile / 100) * (sortedValues.length - 1);
    const lower = Math.floor(index);
    const upper = Math.ceil(index);
    if (lower === upper) return sortedValues[lower];
    return (
      sortedValues[lower] +
      (sortedValues[upper] - sortedValues[lower]) * (index - lower)
    );
  }

  private calculateMAD(values: number[], median: number): number {
    const deviations = values.map((v) => Math.abs(v - median));
    return this.calculatePercentile(
      [...deviations].sort((a, b) => a - b),
      50,
    );
  }

  private metricToAnomalyType(metric: string): AnomalyType {
    if (metric.includes("cost")) return "cost_spike";
    if (metric.includes("latency")) return "latency_anomaly";
    if (metric.includes("error")) return "error_rate_spike";
    if (metric.includes("quality")) return "quality_degradation";
    if (metric.includes("token")) return "token_usage_spike";
    return "custom";
  }

  private calculateSeverity(
    deviation: number,
    _config: AnomalyMetricConfig,
  ): AnomalySeverity {
    const adjustedDeviation = deviation * (1 + this.config.sensitivity * 0.5);

    if (adjustedDeviation > 5) return "critical";
    if (adjustedDeviation > 3) return "warning";
    return "info";
  }

  private calculateConfidence(sampleSize: number, deviation: number): number {
    // Higher confidence with more samples and larger deviations
    const sampleFactor = Math.min(sampleSize / 100, 1);
    const deviationFactor = Math.min(deviation / 5, 1);
    return Math.min(0.5 + sampleFactor * 0.3 + deviationFactor * 0.2, 1);
  }

  private generateSuggestions(type: AnomalyType, _metric: string): string[] {
    const suggestions: Record<AnomalyType, string[]> = {
      cost_spike: [
        "Check for increased traffic or usage",
        "Review recent prompt changes for token efficiency",
        "Investigate if a specific feature/user is responsible",
      ],
      latency_anomaly: [
        "Check LLM provider status page",
        "Review if prompts have increased in size",
        "Check for network issues or rate limiting",
      ],
      error_rate_spike: [
        "Review error logs for common patterns",
        "Check API key validity and rate limits",
        "Investigate if specific tools are failing",
      ],
      quality_degradation: [
        "Review recent prompt changes",
        "Check if model has been updated",
        "Analyze low-scoring responses for patterns",
      ],
      token_usage_spike: [
        "Review prompt templates for efficiency",
        "Check for repeated or redundant context",
        "Investigate if specific features are affected",
      ],
      hallucination_cluster: [
        "Review context and grounding data",
        "Consider adding fact-checking tools",
        "Analyze affected prompts for patterns",
      ],
      tool_failure_pattern: [
        "Check tool service availability",
        "Review tool input patterns for errors",
        "Consider adding retry logic or fallbacks",
      ],
      custom: [
        "Review custom rule configuration",
        "Investigate the specific metric pattern",
      ],
    };

    return suggestions[type] || suggestions.custom;
  }
}

File: packages/sdk-python/src/agentops/anomaly/detector.py
"""
AgentOps SDK - Anomaly Detector

Statistical anomaly detection for LLM operations.
"""

import math
import time
import uuid
from collections import deque
from typing import Dict, List, Optional
from .types import (
    AnomalyConfig,
    MetricSnapshot,
    DetectedAnomaly,
    AnomalyType,
    AnomalySeverity,
    MetricTimeSeries,
)


class AnomalyDetector:
    """
    Real-time anomaly detection for LLM operations.
    
    Uses statistical methods (z-score, threshold) to detect anomalies
    in latency, token usage, error rates, and costs.
    """

    def __init__(self, config: Optional[AnomalyConfig] = None):
        self._config = config or AnomalyConfig()
        self._metrics: Dict[str, deque] = {
            "latency": deque(maxlen=self._config.baseline_window_size),
            "tokens_input": deque(maxlen=self._config.baseline_window_size),
            "tokens_output": deque(maxlen=self._config.baseline_window_size),
            "cost": deque(maxlen=self._config.baseline_window_size),
            "quality": deque(maxlen=self._config.baseline_window_size),
            "error_rate": deque(maxlen=self._config.baseline_window_size),
        }
        self._active_anomalies: Dict[str, DetectedAnomaly] = {}
        self._anomaly_history: List[DetectedAnomaly] = []
        self._error_count = 0
        self._total_count = 0

    @property
    def is_enabled(self) -> bool:
        """Check if anomaly detection is enabled."""
        return self._config.enabled

    def record_metrics(self, snapshot: MetricSnapshot) -> List[DetectedAnomaly]:
        """
        Record a metric snapshot and check for anomalies.
        
        Args:
            snapshot: The metric snapshot to record
            
        Returns:
            List of newly detected anomalies
        """
        if not self._config.enabled:
            return []

        detected = []
        self._total_count += 1
        
        if snapshot.error_occurred:
            self._error_count += 1

        # Record metrics
        if snapshot.latency_ms is not None:
            self._metrics["latency"].append({
                "value": snapshot.latency_ms,
                "timestamp": snapshot.timestamp,
            })
            anomaly = self._check_metric(
                "latency",
                snapshot.latency_ms,
                snapshot.timestamp,
                AnomalyType.LATENCY_SPIKE,
                threshold=self._config.latency_threshold_ms,
            )
            if anomaly:
                detected.append(anomaly)

        if snapshot.tokens_input is not None:
            self._metrics["tokens_input"].append({
                "value": snapshot.tokens_input,
                "timestamp": snapshot.timestamp,
            })
            anomaly = self._check_metric(
                "tokens_input",
                snapshot.tokens_input,
                snapshot.timestamp,
                AnomalyType.TOKEN_EXPLOSION,
                threshold=self._config.token_threshold,
            )
            if anomaly:
                detected.append(anomaly)

        if snapshot.tokens_output is not None:
            self._metrics["tokens_output"].append({
                "value": snapshot.tokens_output,
                "timestamp": snapshot.timestamp,
            })
            anomaly = self._check_metric(
                "tokens_output",
                snapshot.tokens_output,
                snapshot.timestamp,
                AnomalyType.TOKEN_EXPLOSION,
                threshold=self._config.token_threshold,
            )
            if anomaly:
                detected.append(anomaly)

        if snapshot.cost is not None:
            self._metrics["cost"].append({
                "value": snapshot.cost,
                "timestamp": snapshot.timestamp,
            })
            anomaly = self._check_metric(
                "cost",
                snapshot.cost,
                snapshot.timestamp,
                AnomalyType.COST_SURGE,
                threshold=self._config.cost_threshold,
            )
            if anomaly:
                detected.append(anomaly)

        if snapshot.quality_score is not None:
            self._metrics["quality"].append({
                "value": snapshot.quality_score,
                "timestamp": snapshot.timestamp,
            })
            # Quality degrades when it drops (invert logic)
            anomaly = self._check_quality_degradation(
                snapshot.quality_score,
                snapshot.timestamp,
            )
            if anomaly:
                detected.append(anomaly)

        # Check error rate
        if self._total_count >= 10:
            error_rate = self._error_count / self._total_count
            self._metrics["error_rate"].append({
                "value": error_rate,
                "timestamp": snapshot.timestamp,
            })
            if error_rate > self._config.error_rate_threshold:
                anomaly = self._create_anomaly(
                    AnomalyType.ERROR_RATE_INCREASE,
                    "error_rate",
                    error_rate,
                    self._config.error_rate_threshold,
                    snapshot.timestamp,
                )
                if anomaly:
                    detected.append(anomaly)

        # Trigger callback
        for anomaly in detected:
            if self._config.on_anomaly:
                self._config.on_anomaly(anomaly)

        return detected

    def get_active_anomalies(self) -> List[DetectedAnomaly]:
        """Get all currently active anomalies."""
        return list(self._active_anomalies.values())

    def resolve_anomaly(self, anomaly_id: str) -> Optional[DetectedAnomaly]:
        """
        Mark an anomaly as resolved.
        
        Args:
            anomaly_id: The ID of the anomaly to resolve
            
        Returns:
            The resolved anomaly, or None if not found
        """
        anomaly = self._active_anomalies.pop(anomaly_id, None)
        if anomaly:
            anomaly.is_active = False
            anomaly.resolved_at = int(time.time() * 1000)
        return anomaly

    def get_anomaly_history(self) -> List[DetectedAnomaly]:
        """Get historical anomalies."""
        return list(self._anomaly_history)

    def get_metric_time_series(
        self,
        metric_name: str,
        start_time: Optional[int] = None,
        end_time: Optional[int] = None,
    ) -> MetricTimeSeries:
        """
        Get time series data for a metric.
        
        Args:
            metric_name: Name of the metric
            start_time: Start timestamp (ms)
            end_time: End timestamp (ms)
            
        Returns:
            Time series data with statistics
        """
        data = list(self._metrics.get(metric_name, []))
        
        # Filter by time range
        if start_time is not None:
            data = [d for d in data if d["timestamp"] >= start_time]
        if end_time is not None:
            data = [d for d in data if d["timestamp"] <= end_time]

        values = [d["value"] for d in data if d["value"] is not None]
        
        statistics = {}
        if values:
            statistics = {
                "mean": sum(values) / len(values),
                "min": min(values),
                "max": max(values),
                "count": len(values),
            }
            if len(values) > 1:
                mean = statistics["mean"]
                variance = sum((x - mean) ** 2 for x in values) / (len(values) - 1)
                statistics["std_dev"] = math.sqrt(variance)
            else:
                statistics["std_dev"] = 0.0

        return MetricTimeSeries(
            metric_name=metric_name,
            data_points=data,
            statistics=statistics,
        )

    def get_baseline(self, metric_name: str) -> Dict[str, float]:
        """
        Get baseline statistics for a metric.
        
        Args:
            metric_name: Name of the metric
            
        Returns:
            Dictionary with mean, std_dev, min, max
        """
        series = self.get_metric_time_series(metric_name)
        return series.statistics

    def clear_baseline(self, metric_name: Optional[str] = None) -> None:
        """
        Clear baseline data.
        
        Args:
            metric_name: Specific metric to clear, or all if None
        """
        if metric_name:
            if metric_name in self._metrics:
                self._metrics[metric_name].clear()
        else:
            for metric in self._metrics.values():
                metric.clear()

    def _check_metric(
        self,
        metric_name: str,
        value: float,
        timestamp: int,
        anomaly_type: AnomalyType,
        threshold: float,
    ) -> Optional[DetectedAnomaly]:
        """Check a metric for anomalies using z-score and threshold."""
        data = self._metrics.get(metric_name, [])
        
        # Need enough data for z-score
        if len(data) < 3:
            # Fall back to threshold check
            if value > threshold:
                return self._create_anomaly(
                    anomaly_type,
                    metric_name,
                    value,
                    threshold,
                    timestamp,
                )
            return None

        # Calculate z-score
        values = [d["value"] for d in data if d["value"] is not None]
        if not values:
            return None
            
        mean = sum(values) / len(values)
        if len(values) > 1:
            variance = sum((x - mean) ** 2 for x in values) / (len(values) - 1)
            std_dev = math.sqrt(variance)
        else:
            std_dev = 0.0

        # Sensitivity adjusts z-score threshold (3.0 at 0.5, 2.0 at 1.0, 4.0 at 0.0)
        z_threshold = 4.0 - (2.0 * self._config.sensitivity)
        
        if std_dev > 0:
            z_score = (value - mean) / std_dev
            if abs(z_score) > z_threshold:
                return self._create_anomaly(
                    anomaly_type,
                    metric_name,
                    value,
                    mean,
                    timestamp,
                    z_score=z_score,
                )

        # Also check hard threshold
        if value > threshold:
            return self._create_anomaly(
                anomaly_type,
                metric_name,
                value,
                threshold,
                timestamp,
            )

        return None

    def _check_quality_degradation(
        self,
        quality: float,
        timestamp: int,
    ) -> Optional[DetectedAnomaly]:
        """Check for quality score degradation."""
        data = self._metrics.get("quality", [])
        
        if len(data) < 5:
            return None

        values = [d["value"] for d in data if d["value"] is not None]
        if not values:
            return None
            
        mean = sum(values) / len(values)
        
        # Quality degradation: significant drop below mean
        threshold = mean * 0.8  # 20% degradation
        if quality < threshold:
            return self._create_anomaly(
                AnomalyType.QUALITY_DEGRADATION,
                "quality",
                quality,
                mean,
                timestamp,
            )
        return None

    def _create_anomaly(
        self,
        anomaly_type: AnomalyType,
        metric_name: str,
        observed: float,
        expected: float,
        timestamp: int,
        z_score: Optional[float] = None,
    ) -> DetectedAnomaly:
        """Create a new anomaly object."""
        anomaly_id = f"an_{uuid.uuid4().hex[:12]}"
        
        deviation = abs(observed - expected)
        if expected != 0:
            deviation_pct = (deviation / abs(expected)) * 100
        else:
            deviation_pct = 100.0

        # Determine severity
        if deviation_pct > 200 or (z_score and abs(z_score) > 4):
            severity = AnomalySeverity.CRITICAL
        elif deviation_pct > 100 or (z_score and abs(z_score) > 3):
            severity = AnomalySeverity.HIGH
        elif deviation_pct > 50 or (z_score and abs(z_score) > 2):
            severity = AnomalySeverity.MEDIUM
        else:
            severity = AnomalySeverity.LOW

        description = self._generate_description(anomaly_type, metric_name, observed, expected)
        suggested_action = self._suggest_action(anomaly_type)

        anomaly = DetectedAnomaly(
            id=anomaly_id,
            type=anomaly_type,
            severity=severity,
            metric_name=metric_name,
            observed_value=observed,
            expected_value=expected,
            deviation=deviation,
            z_score=z_score,
            detected_at=timestamp,
            is_active=True,
            description=description,
            suggested_action=suggested_action,
        )

        self._active_anomalies[anomaly_id] = anomaly
        self._anomaly_history.append(anomaly)

        return anomaly

    def _generate_description(
        self,
        anomaly_type: AnomalyType,
        metric_name: str,
        observed: float,
        expected: float,
    ) -> str:
        """Generate a human-readable description."""
        descriptions = {
            AnomalyType.LATENCY_SPIKE: f"Latency spike detected: {observed:.1f}ms vs expected {expected:.1f}ms",
            AnomalyType.TOKEN_EXPLOSION: f"Token usage anomaly in {metric_name}: {observed:.0f} vs expected {expected:.0f}",
            AnomalyType.ERROR_RATE_INCREASE: f"Error rate increased to {observed*100:.1f}% (threshold: {expected*100:.1f}%)",
            AnomalyType.COST_SURGE: f"Cost anomaly: ${observed:.4f} vs expected ${expected:.4f}",
            AnomalyType.QUALITY_DEGRADATION: f"Quality degraded to {observed:.2f} from baseline {expected:.2f}",
            AnomalyType.PATTERN_CHANGE: f"Pattern change detected in {metric_name}",
        }
        return descriptions.get(anomaly_type, f"Anomaly in {metric_name}")

    def _suggest_action(self, anomaly_type: AnomalyType) -> str:
        """Suggest remediation action."""
        actions = {
            AnomalyType.LATENCY_SPIKE: "Check for rate limiting, network issues, or model overload",
            AnomalyType.TOKEN_EXPLOSION: "Review prompt length, check for prompt injection, verify context limits",
            AnomalyType.ERROR_RATE_INCREASE: "Check API status, review error logs, verify input validation",
            AnomalyType.COST_SURGE: "Review token usage, check for runaway processes, verify rate limits",
            AnomalyType.QUALITY_DEGRADATION: "Review prompt quality, check model responses, consider prompt refinement",
            AnomalyType.PATTERN_CHANGE: "Investigate recent changes, review data patterns",
        }
        return actions.get(anomaly_type, "Investigate the anomaly")
